src/assets/src/changes.ts

```diff
@@ -0,0 +1,90 @@
+import xorWith from "lodash.xorwith";
+import isEqual from "lodash.isequal";
+
+import { Base, isMeeting, isQueueBase, isUser, Meeting, MeetingStatus, QueueBase } from "./models"
+
+
+export interface ChangeEvent {
+    eventID: number;
+    text: string;
+}
+
+const queueBasePropsToWatch: (keyof QueueBase)[] = ['status', 'name'];
+const meetingPropsToWatch: (keyof Meeting)[] = ['backend_type', 'assignee'];
+
+interface HumanReadableMap {
+    [key: string]: string;
+}
+
+const propertyMap: HumanReadableMap = {
+    'backend_type': 'meeting type',
+    'assignee': 'host'
+}
+
+function detectChanges<T extends Base>(versOne: T, versTwo: T, propsToWatch: (keyof T)[]): string | undefined {
+    for (const property of propsToWatch) {
+        let valueOne = versOne[property] as T[keyof T] | string;
+        let valueTwo = versTwo[property] as T[keyof T] | string;
+        // Check for nested user objects and falsy values
+        if (isUser(valueOne)) valueOne = valueOne.username;
+        if (!valueOne) valueOne = 'None';
+        if (isUser(valueTwo)) valueTwo = valueTwo.username;
+        if (!valueTwo) valueTwo = 'None';
+        if (valueOne !== valueTwo) {
+            // Make some property strings more human readable
+            const propName = (property in propertyMap) ? propertyMap[property as string] : property;
+            return `The ${propName} changed from "${valueOne}" to "${valueTwo}".`;
+        }
+    }
+    return;
+}
+
+
+// https://lodash.com/docs/4.17.15#xorWith
+
+export function compareEntities<T extends Base> (oldOnes: T[], newOnes: T[]): string | undefined
```

Comment:
`Base` can be a lot of things, but only the following types are supported in this function: `Meeting`, `Queue`. You could tighten up the typing here by further constraining your generic constraint to `T extends Manage|QueueBase`.

Reply:
You could also say `type ComparableEntity = Meeting | QueueBase` then make this function non-generic, replacing `T` with `ComparableEntity`. The outcome would be the same. When I see a function that uses generic types, my assumption is that it doesn't make any attempt to figure out exactly what you passed in (IOW, as written I'd expect this function to care about properties in `Base` and nothing else).

Reply:
The other direction you could go in would by making these functions even more generic, factoring out the type checks somehow. I find that recusively comparing JS objects to diff them is a problem I run into surprisingly often, which makes me wonder if there are libraries. A cursory google search shows there are some, but I haven't tried them yet. They might be worth considering. Otherwise, if they don't look appropriate or it would take too much work to pay off, I think this general approach is OK.

Reply:
Nice points, let me think this over and try out the changes you suggest.

Reply:
As far as packages, I've found:
* https://github.com/flitbit/diff
* https://github.com/cosmicanant/recursive-diff

Did you see any others? The first looks fairly reliable, though it hasn't been updated since 2018.
The second is much less widely used.

Reply:
> You could also say `type ComparableEntity = Meeting | QueueBase` then make this function non-generic, replacing `T` with `ComparableEntity`. The outcome would be the same.

I might be betraying my hazy understanding of generics, but question: is there any merit to keeping the generic because otherwise the function could allow `oldOnes` and `newOnes` to be different types of objects if their types are each `ComparableEntity[]`?

Reply:
Hard to say which lib is best without doing a deep dive, but I tend to gravitate toward things that have more uses, stars, contributors, etc. and `flitbit/diff` seems to have quite a lead there. Its types seem to be in `@types/deep-diff`. 

>is there any merit to keeping the generic because otherwise the function could allow oldOnes and newOnes to be different types

Absolutely, good point, keeping it generic makes sure both params are the same type.

Reply:
Hard to say which lib is best without doing a deep dive, but I tend to gravitate toward things that have more uses, stars, contributors, etc. and `flitbit/diff` seems to have quite a lead there. Its types seem to be in `@types/deep-diff`.

> is there any merit to keeping the generic because otherwise the function could allow oldOnes and newOnes to be different types

Absolutely, good point, keeping it generic makes sure both params are the same type.

Reply:
Ah, I must have missed something regarding TypeScript and `deep-diff`. That looks promising; I will investigate that further. `deep-object-diff` was a little unpredictable, and wasn't giving me enough info.

Regarding your `ComparableEntity` suggestion and generics, see commit https://github.com/tl-its-umich-edu/remote-office-hours-queue/pull/258/commits/1bd1d2049a12b02e0afe53df5f52c64325e85405